mitiq/zne/inference.py

```diff
@@ -397,12 +378,183 @@ def is_converged(self) -> bool:
+                for scale, shots in zip(self._scale_factors, self._shot_list)
+            ]
+        else:
+            self._instack = [
+                {"scale_factor": scale} for scale in self._scale_factors
+            ]
+
+    def run_batched(
+        self,
+        qp: QPROGRAM,
+        batched_executor: Callable[..., List[float]],
+        scale_noise: Callable[[QPROGRAM, float], QPROGRAM],
+        num_to_average: int = 1,
+    ) -> "BatchedFactory":
+        """Computes the expectation values at each scale factor by calling
+        `batched_executor` on a list of all circuits to run.
+
+        Args:
+            qp: Quantum circuit to run.
+            batched_executor: Function which inputs a list of circuits and
+                outputs a list of expectation values.
+            scale_noise: Noise scaling function.
+            scale_noise: Noise scaling function.
+            num_to_average: Number of times to call scale_noise at each scale
+                factor.
+        """
+        self.reset()
+        self._batch_populate_instack()
+        kwargs = self._get_keyword_args()
+
+        # Get all noise-scaled circuits to run
+        to_run = self._generate_circuits(qp, scale_noise, num_to_average)
+
+        # Run the circuits in a batch
+        # TODO: Deal with keyword args better.
+        res = batched_executor(to_run, kwargs=kwargs)
+
+        # Average the expectation results
+        self._outstack = [
+            np.average(res[i * num_to_average: (i + 1) * num_to_average])
+            for i in range(len(res) // num_to_average)
+        ]
+
+        return self
+
+    def run(
+        self,
+        qp: QPROGRAM,
+        executor: Tuple[Callable[..., float], Callable[..., List[float]]],
+        scale_noise: Callable[[QPROGRAM, float], QPROGRAM],
+        num_to_average: int = 1,
+    ) -> "BatchedFactory":
+        """Computes the expectation values at each scale factor and stores them
+        in the factory. If the executor returns a single expectation value, the
+        circuits are run sequentially. If the executor is batched and returns
+        a list of expectation values (one for each circuit), then the circuits
+        are sent to the backend as a single job. To detect if an executor is
+        batched, it must be annotated.
+
+        Args:
+            qp: Quantum circuit to run.
+            executor: A "single executor" (1) or a "batched executor" (2).
+                (1) A function which inputs a single circuit and outputs a
+                single expectation value of interest.
+                (2) A function which inputs a list of circuits and outputs a
+                list of expectation values (one for each circuit).
+            scale_noise: Noise scaling function.
+            num_to_average: Number of times to call scale_noise at each scale
+                factor.
+        """
+        self.reset()
+        executor_annotation = inspect.getfullargspec(executor).annotations
```

Comment:
Should we somehow notify or warn the user if they are using `run_batched` but this annotation is empty? So that they know we're defaulting to non-batched?

Reply:
I would say no, since:

1. If they call `run` with a non-annotated executor, `run_batched` will never be called.
1. If they call `run_batched` with a non-annotated but batched executor, we will throw a warning but shouldn't.
1. If they call `run_batched` with a non-annotated, non-batched executor, I'm willing to call that user error.